Document implementation support for selected feature flags (#8)
t-bast/lightning-docs: feature-flags.md

> ## `data_loss_protect`

Reviewer: Can you add a small header paragraph:

```
# Lightning feature flags: from Zero to Hero

This article describes optional lightning features that nodes can turn on and off.
Nodes actively tell the whole network about the features they support, to help peers looking for specific features connect to them.
```

> ## `data_loss_protect`

Reviewer: Can you add a small, high-level description of the feature?

Accepted in November 2017: [meeting log](https://docs.google.com/document/d/1x2Qr2VgqauFfPgs-uBv7B4GegifAzRmfA_LpebvsETM/edit)

This option is required by most implementations, e.g. LND: [brontide.go:2749](https://github.com/lightningnetwork/lnd/blob/cc0321d1881ed23c9608cf898af2e2a7b347304a/peer/brontide.go#L2749).

## `static_remote_key`

> This removes key-rotation from the `to_remote` output. E.g. watchtowers are easier to implement this way.

Reviewer: I would also add that it makes it easier to recover most of your funds in case you lost your channels backup.

- Electrum requires this for all peers since [4.0.1 from July 2020](https://github.com/spesmilo/electrum/blob/922a48f2b74bd6c58682f9d5b6163c4fbd45981a/RELEASE-NOTES#L107)

## `payment_secret`

Merged in July 2019 as part of MPP: https://github.com/lightningnetwork/lightning-rfc/pull/643

The feature prevents probing attacks as noted in BOLT-11.

You can use payment_secret without necessarily using MPP, though it is a requirement for MPP.

### Implementations

- C-Lightning: since at least [v0.8.0 from December 2019](https://github.com/ElementsProject/lightning/releases/tag/v0.8.0), since that version supports receiving MPP
- Eclair: since at least [this PR from October 2019](https://github.com/ACINQ/eclair/pull/1153), released with MPP (see below)
- Electrum: sent/accepted since [4.0.1 from July 2020](https://github.com/spesmilo/electrum/blob/922a48f2b74bd6c58682f9d5b6163c4fbd45981a/RELEASE-NOTES#L107)
- LND: supported since v0.10.0 (see below), **required** in incoming payments since [v0.12.0](https://github.com/lightningnetwork/lnd/pull/4752) (unreleased as of January 7th, 2020):  All generated invoices will include the `s` field.

## Multi-part payments (MPP)

> Contrast with e.g. "High" Atomic Multi-part Payments (AMP), which is different, not yet standardized, but mentioned in BOLT-04.

Reviewer: I think this sentence isn't useful at all for newcomers, it's even more confusing!
Why not simply explain that this is a feature to split payments into multiple chunks to ensure they're more likely to find routes?

### Implementations
All listed implementations that have send support, also have receive support.

- C-Lightning: Send support since [v0.9.0 from July 2020](https://github.com/ElementsProject/lightning/releases/tag/v0.9.0)

> - Eclair: Send support since [v0.3.3 from January 2020](https://github.com/ACINQ/eclair/releases/tag/v0.3.3), see also [wiki page](https://github.com/ACINQ/eclair/wiki/Multipart-Payments). Needs manual enabling.

Reviewer: ```suggestion
- Eclair: Send support since [v0.3.3 from January 2020](https://github.com/ACINQ/eclair/releases/tag/v0.3.3), see also [wiki page](https://github.com/ACINQ/eclair/wiki/Multipart-Payments).
```

I've updated the wiki page, we turned this on by default a while ago (for both sending and receiving).

# Lightning feature flags: from Zero to Hero

This article describes optional lightning features that nodes can turn on and off.
Nodes actively tell the whole network about the features they support, to help peers looking for specific features connect to them.

## `data_loss_protect`

> This flag allows a node to detect upon channel reestablishment that it has lost state (i.e. thinks a state is 'current' while it has been superseded and maybe even revoked). This works because the commitment point (that is added with this flag) can be verified to be authenticly derived from the secret, so the 'forgetful' node must have sent it to them before it lost its commitment point counter state.

Reviewer: nit:

```suggestion
This flag allows a node to detect upon channel reestablishment that it has lost state (i.e. thinks a state is 'current' while it has been superseded and maybe even revoked). This works because the commitment point (that is added with this flag) can be verified to be authentically derived from the secret, so the 'forgetful' node must have sent it to them before it lost its commitment point counter state.
```

- Electrum requires this for all peers since [4.0.1 from July 2020](https://github.com/spesmilo/electrum/blob/922a48f2b74bd6c58682f9d5b6163c4fbd45981a/RELEASE-NOTES#L107)

## `payment_secret`

Merged in July 2019 as part of MPP: https://github.com/lightningnetwork/lightning-rfc/pull/643

The feature prevents probing attacks as noted in BOLT-11.

You can use payment_secret without necessarily using MPP, though it is a requirement for MPP.

### Implementations

- C-Lightning: since at least [v0.8.0 from December 2019](https://github.com/ElementsProject/lightning/releases/tag/v0.8.0), since that version supports receiving MPP
- Eclair: since at least [this PR from October 2019](https://github.com/ACINQ/eclair/pull/1153), released with MPP (see below)
- Electrum: sent/accepted since [4.0.1 from July 2020](https://github.com/spesmilo/electrum/blob/922a48f2b74bd6c58682f9d5b6163c4fbd45981a/RELEASE-NOTES#L107)
- LND: supported since v0.10.0 (see below), **required** in incoming payments since [v0.12.0](https://github.com/lightningnetwork/lnd/pull/4752) (unreleased as of January 7th, 2020):  All generated invoices will include the `s` field.

## Multi-part payments (MPP)

> This splits a payment into multiple pieces, that all have the same payment hash (so it is [vulnerable to a straddling attack](https://lists.linuxfoundation.org/pipermail/lightning-dev/2018-February/000993.html)). This means that the theoretical maximum payment amount can be higher. A receiver doesn't know how many parts a payment is split into. To avoid an attack where one part the MPP payment will never arrive and the HTLC consumes resources over too long, LND has a [two-minute timeout heuristic](https://lightning.engineering/posts/2020-05-07-mpp/), after which all HTLCs are canceled.

Reviewer: The timeout is provided by the spec, not an lnd implementation detail (see Bolt 4):

```suggestion
This splits a payment into multiple pieces, that all have the same payment hash (so it is [vulnerable to a straddling attack](https://lists.linuxfoundation.org/pipermail/lightning-dev/2018-February/000993.html)). This means that the theoretical maximum payment amount can be higher. A receiver doesn't know how many parts a payment is split into. To avoid an attack where one part the MPP payment will never arrive and the HTLC consumes resources over too long, receivers must fail incoming HTLCs after a reasonable timeout, usually 60 seconds.
```
